codeworld-server/src/Collaboration_.hs

```diff
@@ -0,0 +1,246 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-
+  Copyright 2017 The CodeWorld Authors. All rights reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-}
+
+module Collaboration_ where
+
+import qualified Control.Concurrent.STM as STM
+import           Control.Monad.State.Strict (StateT)
+import           Control.Monad.Trans
+import           Control.Monad.Trans.Reader (ReaderT)
+import qualified Control.OperationalTransformation.Selection as Sel
+import qualified Control.OperationalTransformation.Server as OTS
+import           Data.Aeson
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Char8 as BC
+import qualified Data.HashMap.Strict as HM
+import           Data.Maybe (fromJust)
+import           Data.Text (Text)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import           Data.Time.Clock
+import qualified Network.SocketIO as SIO
+import           Snap.Core
+import           System.FilePath
+
+import CollaborationUtil
+import DataUtil
+import Model
+import SnapUtil
+
+collabRoutes :: Snap () -> ClientId -> [(B.ByteString, Snap ())]
+collabRoutes socketIOHandler clientId =
+    [ ("addToCollaborate",  addToCollaborateHandler clientId)
+    , ("collabShare",       collabShareHandler clientId)
+    , ("listCurrentOwners", listCurrentOwnersHandler clientId)
+    , ("socket.io", socketIOHandler)
+    ]
+
+getFrequentParams :: Int -> ClientId -> Snap (User, BuildMode, FilePath)
+getFrequentParams getType clientId = do
+    user <- getUser clientId
+    mode <- getBuildMode
+    case getType of
+        1 -> do
```

Comment:
Please, no!  At least, use the type system.  This is Haskell.  But better yet, decompose the problem in a way that's less arbitrary.